Peer details: replace Direction with Connection Type

jonatack opened the pull request:
![Screenshot from 2021-01-09 11-23-17](https://user-images.githubusercontent.com/2415484/104089297-c5e18d80-5265-11eb-9251-49afcfdb562b.png)

Closes #159.

MarcoFalke commented:
Nice. Concept ACK

MarcoFalke commented on src/qt/guiutil.cpp:
```diff
@@ -749,6 +749,20 @@ QString boostPathToQString(const fs::path &path)
+    // Inbound peer connections
+    if (conn_type == "inbound") return "Inbound full relay";
+    if (conn_type == "inbound-block-relay") return "Inbound block relay";
+    return "";
```
nit: Would be nice if this default fallback was replaced with an assert(false) and the above ifs with an enum-switch-case

jonatack replied:
Agree, but ISTM that only works for integer types.

MarcoFalke replied:
`ConnectionType` is an enum class (implementation defined integer)

jonatack replied:
Right, I wanted to do that but the `ConnectionType` enum wasn't exposed in CNodeStats. Ok, done.

hebasto commented:
Concept ACK.

hebasto commented on src/qt/guiutil.h:
```diff
@@ -224,6 +225,9 @@ namespace GUIUtil
     /* Convert OS specific boost path to QString through UTF-8 */
     QString boostPathToQString(const fs::path &path);
 
+    /* Convert ConnectionType enum to QString */
```
571b0469fa952be98874d1afd772a7fa708dd83f

This comment is not Doxygen-compatible.

jonatack replied:
thanks! fixed

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -749,6 +749,19 @@ QString boostPathToQString(const fs::path &path)
     return QString::fromStdString(path.string());
 }
 
+QString ConnectionTypeToQString(ConnectionType conn_type)
+{
+    switch (conn_type) {
+    case ConnectionType::INBOUND: return "Inbound full relay";
+    case ConnectionType::OUTBOUND_FULL_RELAY: return "Outbound full relay";
+    case ConnectionType::BLOCK_RELAY: return "Outbound block relay";
+    case ConnectionType::MANUAL: return "Outbound manual";
+    case ConnectionType::FEELER: return "Outbound feeler";
+    case ConnectionType::ADDR_FETCH: return "Outbound address fetch";
+    } // no default case, so the compiler can warn about missing cases
+    assert(false);
+}
+
```
571b0469fa952be98874d1afd772a7fa708dd83f

The returned values should be translated.

UX note: I have a personal preference do not capitalize the first letter of returned values. Feel free to ignore it.

jonatack replied:
I also prefer all lowercase (e.g. rpc/cli) but the GUI UI uses capitalization (I hesitated to use titlecase here).

hebasto replied:
I perceive it like "Connection type: Outbound full relay". That's why I prefer the lowercased variant.

If the capitalized variant will be required in some other places in the GUI, it is easy to provide it with a generic function.

I'm not a design expert, then I'm fine with any variant.

jonatack replied:
Thanks--translated and changed to titlecase for coherence with the general GUI UI.

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -749,6 +749,19 @@ QString boostPathToQString(const fs::path &path)
+QString ConnectionTypeToQString(ConnectionType conn_type)
+{
+    switch (conn_type) {
+    case ConnectionType::INBOUND: return "Inbound full relay";
```
571b0469fa952be98874d1afd772a7fa708dd83f

It seems not always true :)

IIUC, you recent changes in `bitcoin-cli` distinguish "block relay" for inbound connections as well.

jonatack replied:
> IIUC, you recent changes in `bitcoin-cli` distinguish "block relay" for inbound connections as well.

Yes, I'm anticipating `inbound full relay` and `inbound block relay` connection types (https://github.com/bitcoin/bitcoin/pull/20726 and https://github.com/bitcoin/bitcoin/pull/20729) but I guess it's better to revert to just inbound for now and update it later.

jonatack replied:
(in the latest -netinfo update in https://github.com/bitcoin/bitcoin/pull/20764, the issue is avoided by using the connection type only to parse full/block/manual/feeler/addr, but not for the direction)

jonatack replied:
In the latest -netinfo update in https://github.com/bitcoin/bitcoin/pull/20764, the issue is avoided by using the connection type only to parse full/block/manual/feeler/addr, but not for the direction. Maybe best to do the same here, hm.

jonatack replied:
Now "Inbound"

hebasto commented on src/qt/forms/debugwindow.ui:
```diff
@@ -1077,14 +1077,17 @@
                <item row="1" column="0">
-                <widget class="QLabel" name="label_23">
+                <widget class="QLabel" name="peerConnectionTypeLabel">
+                 <property name="toolTip">
+                  <string>The connection type of this peer.</string>
+                 </property>
```
bf8135db2a658aed3129ec82e237f8030a8ec7dd

I've noted your [comment](https://github.com/bitcoin-core/gui/pull/162#discussion_r548865476), but here the tooltip provides no additional info. I don't think that adding such a tooltip for the sake itself is a good approach for the GUI:
- https://uxplanet.org/tooltips-in-ui-design-f63e117aa3d1
- https://www.nngroup.com/articles/tooltip-guidelines/

jonatack replied:
True. Perhaps it should link to more information about the connection types (like we have in the codebase), maybe in a markdown file in `/doc`.

jonatack replied:
Maybe I should add more useful info, like in the RPC help (getpeerinfo)

hebasto replied:
Anyway, please keep it useful and _succinct_, or else do not add it at all :)

jonatack replied:
Added info similar to the getpeerinfo help, see updated screenshot in the PR description.

hebasto reviewed (changes requested):
Concept ACK.

I have reviewed the code (bf8135db2a658aed3129ec82e237f8030a8ec7dd), and tested it.

jonatack commented:
Thanks for the excellent feedback @hebasto! Updated.

hebasto commented:
I have reviewed the code (8e7f82a6a84c7b2553f88767dd356f480d1ae4b7), and tested it.

The following patch:
```diff
--- a/src/qt/forms/debugwindow.ui
+++ b/src/qt/forms/debugwindow.ui
@@ -1079,7 +1079,7 @@
                <item row="1" column="0">
                 <widget class="QLabel" name="peerConnectionTypeLabel">
                  <property name="toolTip">
-                  <string>The type of peer connection: Outbound Full Relay (default), Outbound Block Relay (does not relay transactions or addresses), Inbound (initiated by peer), Outbound Manual (added using RPC addnode or -addnode/-connect config options), Outbound Feeler (short-lived, for testing addresses), or Outbound Address Fetch (short-lived, for soliciting addresses).</string>
+                  <string>The type of peer connection: Outbound Full Relay (default), Outbound Block Relay (does not relay transactions or addresses), Inbound (initiated by peer), Outbound Manual (added using RPC %1 or %2/%3 config options), Outbound Feeler (short-lived, for testing addresses), or Outbound Address Fetch (short-lived, for soliciting addresses).</string>
                  </property>
                  <property name="text">
                   <string>Connection Type</string>
--- a/src/qt/rpcconsole.cpp
+++ b/src/qt/rpcconsole.cpp
@@ -496,6 +496,8 @@ RPCConsole::RPCConsole(interfaces::Node& node, const PlatformStyle *_platformSty
     clear();
 
     GUIUtil::handleCloseWindowShortcut(this);
+
+    ui->peerConnectionTypeLabel->setToolTip(ui->peerConnectionTypeLabel->toolTip().arg("addnode").arg(QString(nonbreaking_hyphen) + "addnode").arg(QString(nonbreaking_hyphen) + "connect"));
 }
 
 RPCConsole::~RPCConsole()
```

- removes RPC and option names from the translatable string
- prevents effects like this
![DeepinScreenshot_select-area_20201225232751](https://user-images.githubusercontent.com/32963518/103142246-d55ec380-4708-11eb-97d2-7e33ede86f98.png)

jonatack commented:
Thanks @hebasto, tested and added your suggestion, updated the screenshot in the PR description.

hebasto reviewed (approved):
ACK 6cd242c50f4a8d3f3c509937d80962fe58096ead, tested on Linux Mint 20 (x86_64) + Qt 5.12.8.

Looks great:
![Screenshot from 2020-12-25 23-52-27](https://user-images.githubusercontent.com/32963518/103142484-78fda300-470c-11eb-9e9e-0aad3a13b24a.png)

promag commented on src/net.h:
```diff
@@ -725,6 +725,7 @@ class CNodeStats
     // Name of the network the peer connected through
     std::string m_network;
     uint32_t m_mapped_as;
+    ConnectionType m_conn_type;
```
a60f67b2fb0eede205db8cce11ad0e6f42dbc90e

This was suggested here https://github.com/bitcoin/bitcoin/pull/19725#discussion_r473361238 and now makes 
`m_conn_type_string` redundant, don't makes sense to keep both TBH. Since `m_conn_type_string` is only used in `getpeerinfo` then drop it and refactor `ConnectionTypeAsString` to take `conn_type` instead?

MarcoFalke replied:
Concept ACK, but RPC changes (refactoring) can only be made in the main repo

jonatack replied:
Done, I think. My initial version of this PR did not touch any non-GUI code. Should I close this PR and re-open in the main repo? (I'm not sure it's worth losing the review above.)

hebasto replied:
Maybe just open a pr in the main repo with the first and last commits?

promag replied:
IMO this isn't a RPC change, it's just a member refactor in a net struct, and the refactor is motivated by this GUI change. Otherwise this can go forward in 2 ways I think, either leave this as it was, and then in main repo drop the redundant member, or 1st refactor and make this dependant of that PR.

jonatack replied:
Yes, this doesn't change the RPC API.

Please let me know which option is best:

1. split it into two pulls, one here, gated by a another one in the other repo
2. move the pull to the other repo
3. revert to the original version that didn't touch main repo code
4. leave this as-is

jonatack replied:
I'm not sure the refactoring changes make sense without the whole. Opened the same pull in https://github.com/bitcoin/bitcoin/pull/20778 and will let the maintainers decide.

promag replied:
Just noting that redundancy is not a blocker, but could be avoided.

jonatack commented:
It was suggested to re-open here.

MarcoFalke commented:
Concept re-ACK. Thanks for re-opening.

jonatack commented:
Rebased.

hebasto reviewed (approved):
re-ACK aef969363a7685fc805833a3ec6863c27500d02f, since my [previous](https://github.com/bitcoin-core/gui/pull/163#pullrequestreview-558838006) review only rebased and the _"p2p, rpc, refactor: remove CNodeStats::m_conn_type_string"_ commit added.

jonatack commented:
Rebased `git range-diff ae8f797 aef9693 7098509`

hebasto commented on src/rpc/net.cpp:
```diff
@@ -249,7 +249,7 @@ static RPCHelpMan getpeerinfo()
         }
         obj.pushKV("bytesrecv_per_msg", recvPerMsgCmd);
-        obj.pushKV("connection_type", stats.m_conn_type_string);
+        obj.pushKV("connection_type", ConnectionTypeAsString(stats.m_conn_type));
```
70985097c00827de94a7922620b63e1ee10dd53d

Mind fixing https://github.com/bitcoin/bitcoin/pull/19725/commits/395acfa83a5436790c1a722a5609ac9d48df235f (https://github.com/bitcoin/bitcoin/pull/19725) by aligning the order of the "connection_type" field both in RPC help and in RPC output?

jonatack replied:
thanks; done

hebasto reviewed (approved):
re-ACK 70985097c00827de94a7922620b63e1ee10dd53d, only rebased since my [previous](https://github.com/bitcoin-core/gui/pull/163#pullrequestreview-560126996) review.

jonatack commented:
Thanks @hebasto, updated the last commit to move the getpeerinfo connection_type help to the correct place per review feedback in https://github.com/bitcoin-core/gui/pull/163#discussion_r550863551 and updated the commit message.

hebasto reviewed (approved):
re-ACK be8294ff428da7d13043b3f9a27d7ace1c2dadf4

laanwj commented:
Concept ACK

jonatack commented:
Rebased after the merge of https://github.com/bitcoin/bitcoin/pull/20786.

hebasto reviewed (approved):
re-ACK 20fa97560b2d27f94f556a570f3556356877b79f, only rebased since my [previous](https://github.com/bitcoin-core/gui/pull/163#pullrequestreview-560723519) review due to the conflict with https://github.com/bitcoin/bitcoin/pull/20786, verified with
```
$ git range-diff master be8294ff428da7d13043b3f9a27d7ace1c2dadf4 20fa97560b2d27f94f556a570f3556356877b79f
```

jarolrod reviewed (commented):
ACK 20fa97560b2d27f94f556a570f3556356877b79f

Tested on macOS Big Sur 11.1 with qt 5.15.2. Looks great!
<img width="497" alt="Screen Shot 2021-01-08 at 11 01 46 AM" src="https://user-images.githubusercontent.com/23396902/104039378-4c359b00-51a4-11eb-827e-8b5f75dc2f20.png">

Small NIT: There's a lot of text in the tooltip. It would be nice if all the possible values were put into a list instead of separated by commas. It would look something like [this](https://ptgmedia.pearsoncmg.com/images/chap17_9780132354165/elementLinks/catalogentry.jpg), which to me would be an improvement.

See: [Supported HTML Subset](https://doc.qt.io/qt-5/richtext-html-subset.html),

jonatack commented:
Better? If yes, bold or no?
![Screenshot from 2021-01-09 00-04-04](https://user-images.githubusercontent.com/2415484/104073609-0c9f9b00-5206-11eb-80ce-5d1979672662.png)
![Screenshot from 2021-01-08 23-57-49](https://user-images.githubusercontent.com/2415484/104073612-0e695e80-5206-11eb-94be-b3f6bff04754.png)

jarolrod commented:
@jonatack the list looks great! I think not-bold is better. If i may, I would suggest one more thing:

Write the descriptions like this:
`
Outbound Block Relay: does not relay transactions or addresses
`

instead of:
`
Outbound Block Relay (does not relay transactions or addresses)
`

jonatack commented:
@jarolrod test with inbound first and your colon suggestion (agree, it's more readable)

![Screenshot from 2021-01-09 01-10-03](https://user-images.githubusercontent.com/2415484/104076853-432de380-520f-11eb-8bd4-59fa1050e401.png)

jarolrod commented:
@jonatack looks awesome 👍

jonatack commented:
Thanks @jarolrod. Updated in latest push, if you'd like to review/test. @hebasto, can you verify the changed tooltip code?

hebasto commented:
> @hebasto, can you verify the changed tooltip code?

The used HTML code is [valid](https://doc.qt.io/qt-5/richtext-html-subset.html).

But I'd prefer to keep the tooltip content in the form `*.ui` file. It makes usage of design tools (e.g., Qt Designer) much easier. In `*.cpp` file should be text substitution code only.

jonatack commented:
@hebasto thanks--done.

hebasto reviewed (approved):
re-ACK 06ba9b300866f33e21512af9d7d2891ee1501bf4, the tooltip content is organized as unordered list.

jarolrod reviewed (commented):
re-ACK 06ba9b300866f33e21512af9d7d2891ee1501bf4

MarcoFalke commented on src/qt/forms/debugwindow.ui:
```diff
@@ -1077,14 +1077,17 @@
-                <widget class="QLabel" name="label_23">
+                <widget class="QLabel" name="peerConnectionTypeLabel">
+                 <property name="toolTip">
+                   <string>The type of peer connection:&lt;ul&gt;&lt;li&gt;Inbound: initiated by peer&lt;/li&gt;&lt;li&gt;Outbound Full Relay: default&lt;/li&gt;&lt;li&gt;Outbound Block Relay: does not relay transactions or addresses&lt;/li&gt;&lt;li&gt;Outbound Manual: added using RPC %1 or %2/%3 configuration options&lt;/li&gt;&lt;li&gt;Outbound Feeler: short-lived, for testing addresses&lt;/li ...
```
translation nit: How are translators supposed to translate `:&lt;ul&gt;&lt;li&gt;`? It might be better to say "Can be one of %s" and then use `Join(translated_strings, ", ")` or simply "Refer to the getpeerinfo RPC help for details on the connection types"?

jonatack replied:
@hebasto, if we replace the escaped markup with placeholders (as done here for the nonbreaking hyphen), is that still QT Designer-friendly? Can do it in #179.

hebasto replied:
> translation nit: How are translators supposed to translate `:&lt;ul&gt;&lt;li&gt;`?

There was a discussion in private IRC with @jonatack:
> \<hebasto\>
> 14:24:52 but I expect the translators' burden :)
> 14:25:39 hope that translators are aware of escaping :)

I'm not an expert in translation but I believe that `:&lt;ul&gt;&lt;li&gt;` is not a subject to translate as `%1` or `%2`.

> It might be better to say "Can be one of %s" and then use `Join(translated_strings, ", ")` or simply "Refer to the getpeerinfo RPC help for details on the connection types"?

Either of the suggested variants LGTM.

> if we replace the escaped markup with placeholders (as done here for the nonbreaking hyphen), is that still QT Designer-friendly?

I think it is. But, tbh, I'd prefer more succinct @MarcoFalke's suggestions.

jonatack replied:
Thanks. Helpfulness and readability for users were the reasons for the iterations above and it seems best to not regress on those improvements with referring GUI users to the command line or dropping the readable list layout. I'll propose replacing the escaped markup with placeholders.

hebasto replied:
Why not a single placeholder for the whole unordered list? We still have a (shorter) tooltip in the form file, and make translators' life easier :)

jonatack replied:
SGTM too.

jonatack replied:
Done in #180
